torch/csrc/autograd/saved_variable.cpp

```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     version_counter_ = impl::version_counter(variable);
     saved_version_ = version_counter_.current_version();
 
+    // This check actually does not hold
+    // TORCH_CHECK(is_output && variable.is_leaf(), "Variable is both an output and a leaf");
+
+    // If the variable is not an output, we can safely save the
+    // original variable without running the risk of reference cycles.
+    // If the variable is not an output, its grad_fn has already been fully
+    // created and in particular will be a different Node than the one
+    // we are currently constructing (the one that owns this SavedVariable).
+    // Similarly, if the variable is a leaf.
```

Comment:
I think the leaf case is different actually. You don't have a cycle because the Tensor only has a weak ref to the grad_accumulator. So that cannot create a cycle.